buffer/src/main/java/io/netty/buffer/ReadableObject.java

```diff
@@ -0,0 +1,166 @@
+     * identical to {@code r.slice(r.readerPosition(), r.readableBytes())}.
+     * This method does not modify {@code readerPosition}.
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the
+     * reference count will NOT be increased.
+     */
+    ReadableObject slice();
+
+    /**
+     * Returns a slice of this object's sub-region. This method does not modify
+     * {@code readerPosition}.
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the
+     * reference count will NOT be increased.
+     *
+     * @param position the starting position for the slice.
+     *
+     * @param length the size of the new slice relative to {@code position}.
+     *
+     * @throws IndexOutOfBoundsException
+     *         if any part of the requested region falls outside of the currently readable region.
+     */
+    ReadableObject slice(long position, long length);
+
+    /**
+     * Returns a new slice of this object's sub-region starting at the current
+     * {@link #readerPosition()} and increases the {@code readerPosition} by the size
+     * of the new slice (= {@code length}).
+     * <p>
+     * Also be aware that this method will NOT call {@link #retain()} and so the
+     * reference count will NOT be increased.
+     *
+     * @param length the size of the new slice
+     *
+     * @return the newly created slice
+     *
+     * @throws IndexOutOfBoundsException
+     *         if {@code length} is greater than {@link #readableBytes()}
+     */
+    ReadableObject readSlice(long length);
+
+    /**
+     * Return the underlying {@link ReadableObject} instance if this is a wrapper around another
+     * object (e.g. a slice of another object).
+     *
+     * @return {@code null} if this is not a wrapper
+     */
+    ReadableObject unwrap();
+
+    /**
+     * Discards the bytes between position 0 and {@code readerPosition}.
+     * <p>
+     * Please refer to the class documentation for more detailed explanation.
+     */
+    ReadableObject discardReadBytes();
+
+    /**
+     * Similar to {@link #discardReadBytes()} except that this method might discard
+     * some, all, or none of read bytes depending on its internal implementation to reduce
+     * overall memory bandwidth consumption at the cost of potentially additional memory
+     * consumption.
+     */
+    ReadableObject discardSomeReadBytes();
+
+    /**
+     * Transfers this object's data to the specified stream starting at the
+     * given {@code readerPosition}. Does not change the {@code readerPosition}.
+     *
+     * @param pos the starting position for the transfer. This must be < {@link #readerLimit()}.
+     * @param length the maximum number of bytes to transfer
+     *
+     * @return the actual number of bytes written out to the specified channel
+     *
+     * @throws IndexOutOfBoundsException
+     *         if {@code length} is greater than {@code this.readableBytes}
```

Comment:
@nmittler - On second though...the statement should also be sensitive to `pos` too right?